mono/metadata/class-init.c

```diff
@@ -3501,6 +3501,41 @@ type_has_references (MonoClass *klass, MonoType *ftype)
 	return FALSE;
 }
 
+static gboolean
+mono_class_is_gparam_with_reference_parent (MonoClass *klass)
+{
+	MonoType *type = m_class_get_byval_arg (klass);
+	g_assert (mono_type_is_generic_parameter (type));
+	MonoGenericParam *gparam = type->data.generic_param;
+	if ((mono_generic_param_info (gparam)->flags & GENERIC_PARAMETER_ATTRIBUTE_REFERENCE_TYPE_CONSTRAINT) != 0)
+		return TRUE;
+	if ((mono_generic_param_info (gparam)->flags & GENERIC_PARAMETER_ATTRIBUTE_VALUE_TYPE_CONSTRAINT) != 0)
+		return FALSE;
+
+	/* FIXME: is this reasonable? */
+	if (mono_generic_param_owner (gparam)->is_anonymous)
+		return FALSE;
```

Comment:
I had to exit early if the generic container is anonymous here in order for the AOT step not to crash - I guess we could be instantiating something with a gshared type.

I think I need to look at the gshared constraint to find out if it's a valuetype or a reference type.